modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -1564,9 +1564,17 @@ else if (!grp.isLocal())
                     log.debug("Notifying exchange future about single message: " + exchFut);
 
                 if (msg.client() && !exchFut.isDone()) {
-                    if (exchFut.initialVersion().compareTo(readyAffinityVersion()) <= 0) {
+                    AffinityTopologyVersion initVer = exchFut.initialVersion();
+                    AffinityTopologyVersion readyVer = readyAffinityVersion();
+
+                    DiscoveryEvent firstEvt = exchFut.firstEvent();
+
+                    if (initVer.compareTo(readyVer) <= 0
+                        && (firstEvt != null && firstEvt.type() != EVT_NODE_LEFT && firstEvt.eventNode().isClient())) {
```

Comment:
Suppose, we missed here EVT_NODE_FAIL event type.

From my point of view we should add new mehtod like this `ExchangeDiscoveryEvents#serverLeftEvent` but for client nodes and use it here. 
Also, I think `firstEvt` always not `null`, so we don't need to check it here but we must sure about it.

Reply:
Good catch with EVT_NODE_FAIL event, I've addressed it.

Null check for ```firstEvt``` is correct here because in case of ExchangeFuture being preempted from exchange queue it is recreated with default value of this field.

I'm not sure I understand your comment about ```serverLeftEvent``` for client nodes. 
This code replies to **client exchange** that was preempted from the queue and thus cannot be finished so client should reconnect.
Additional checks were added to prevent situations when client nodes may be forced to reconnect when server nodes leave topology.